### PR TITLE
Gracefully handle None-valued tag fields

### LINKED ISSUE #3546: [BUG] Clearing the tags field and then tagging samples raises error
To reproduce:
```python
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart").clone()
dataset.clear_sample_field("tags")
dataset.tag_samples("test")

# ValueError: Cannot apply $addToSet to non-array field. Field named 'tags' has non-array type null
```

This is happening because `clear_sample_field` sets the values of the field to `None` for each sample, but `tag_samples()` then tries to add a value to what it expects is a list.

We should probably add a check to `tag_samples()` to verify that the field isn't `None` before adding samples. Alternatively, we could have `clear_sample_field("tags")` check if you are clearing the tags field and set it to `[]` instead of `None`, but that field could still be set to `None` in other ways, so it seems the first approach would be preferred.

### PR DESCRIPTION
**Pull request by brimoor**
## Change log

Resolves #3546 

By convention, all non-required FO fields should be nullable, but the implementation of `tag_samples()`, `untag_samples()`, `tag_labels()`, and `untag_labels()` use the `$addToSet` operator, which gracefully handles missing fields, but unfortunately cannot handle `null` fields. So if a user clears their tags via `clear_sample_field("tags")`, for example, then all tagging operations will start to fail.

Note that the default behavior is to set `tags = []`, so having `tags` fields that contain `None` is likely **rare**.

## Implementation choice

Why use `try-except`? I considered two alternatives:

1. Refactor methods like `tag_samples()` to use two separate batch edits: one for documents whose `tags` field contains a list, and a separate one for documents whose `tags` field is missing/null:

```py
view1 = self.exists("tags").match_tags(tags, bool=False, all=True)
view1._edit_sample_tags({"$addToSet": {"tags": {"$each": list(tags)}}})

view2 = self.exists("tags", bool=False)
view2._edit_sample_tags({"$set": {"tags": tags}})
```

The drawback here is that this *always* requires two full passes over the samples collection, which would roughly halve performance just to handle the null case, which I assume is **rare**.

2. Encode the conditional operation in the update operation. Unfortunately, `$addToSet` is not available in aggregation pipelines, so the only way to achieve this would be something like this:

```py
# current
update = {"$addToSet": {"tags": {"$each": tags}}}

# new aggregation pipeline-style
pipeline = [
    {
        "$cond": {
            "if": {"$gt": ["$tags", None]},
            "then": {"$setUnion": ["$tags", tags]},
            "else": tags,
        },
    }
]
```

Unfortunately, `$setUnion` is not a 1-1 replacement for `$addToSet`: it will silently remove existing duplicates and makes no guarantee that the ordering of the existing tags will stay the same. So, really a non-starter.

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved error handling in tagging functionalities to address issues with null values in tags, ensuring more robust data operations.

- **Documentation**
	- Updated method descriptions in the dataset module for clarity and accuracy.

- **Tests**
	- Added tests to verify the robustness of the tagging functionalities when encountering null values.

### CODE REVIEW

File: fiftyone/core/collections.py [hunk 1]
```diff
@@ -1680,9 +1700,31 @@
             tags = list(tags)
             update_fcn = lambda path: {"$addToSet": {path: {"$each": tags}}}
```
[the remark this answers is not in the document]
Comment (swheaton) [reply fiftyone/core/collections.py:1701]: In this case, the lambda is sufficiently simple that it's not too bad for readability.

File: fiftyone/core/collections.py [hunk 2]
```diff
@@ -1680,9 +1700,31 @@
...
+            # Note that its okay to run $addToSet multiple times as the tag
+            # won't be added multiple times.
+            #
+            # For future reference, the error message looks roughly like this:
+            #   ValueError: Cannot apply $addToSet to non-array field. Field
+            #               named 'tags' has non-array type null
+            #
+            if "null" in str(e):
+                _, tags_path = self._get_label_field_path(label_field, "tags")
+                none_tags = self.filter_labels(label_field, F("tags") == None)
```
[the remark this answers is not in the document]
Comment (swheaton) [reply fiftyone/core/collections.py:1720]: Usually you'd be correct. However in this case `F("tags")` is the start of a `ViewExpression` which is a complex grammar defined elsewhere in this repository. `== None` is a valid operator on a `ViewExpression` while `is None` is not. Usually this class is aliased as `F` and imported like this: `from fiftyone import ViewExpression as F`

File: fiftyone/core/dataset.py [hunk 3]
```diff
@@ -1778,7 +1778,7 @@ def clear_sample_field(self, field_name):
         The field will remain in the dataset's schema, and all samples will
         have the value ``None`` for the field.
 
-        You can use dot notation (``embedded.field.name``) to clone embedded
+        You can use dot notation (``embedded.field.name``) to clear embedded
         frame fields.
```
Comment (swheaton) [inline fiftyone/core/dataset.py:1782]: Good eye but now I think the word "frame" is another copy pasta accident.
```suggestion
        You can use dot notation (``embedded.field.name``) to clear embedded
        sample fields.
```

File: fiftyone/core/dataset.py [hunk 4]
```diff
@@ -1792,7 +1792,7 @@ def clear_sample_fields(self, field_names):
...
```
Comment (swheaton) [inline fiftyone/core/dataset.py:1796]:
```suggestion
        You can use dot notation (``embedded.field.name``) to clear embedded
        sample fields.
```

**Review by swheaton — APPROVED** [review]
lgtm; thank for adding tests.
minor suggestions for further de-copypasta-ing.

Another suggestion is if we can get the original author in the commit history somehow for contribution credit.

**Comment by brimoor** [comment]
> minor suggestions for further de-copypasta-ing.

@swheaton thanks for catching! 🙇

**Pull request merged**